runners/apex/src/main/java/org/apache/beam/runners/apex/translation/TranslationContext.java

```diff
@@ -191,11 +209,74 @@ public void populateDAG(DAG dag) {
     }
   }
 
+  private void optimizeStreams(DAG.StreamMeta streamMeta,
+                               Map.Entry<PCollection,
+                                         Pair<OutputPortInfo, List<InputPortInfo>>> streamEntry) {
+    DAG.Locality loc = null;
+
+    List<InputPortInfo> sinks = streamEntry.getValue().getRight();
+    OutputPortInfo source = streamEntry.getValue().getLeft();
+    PTransform sourceTransform = source.transform.getTransform();
+    if (sourceTransform instanceof ParDo.Bound
+        || sourceTransform instanceof ParDo.BoundMulti) {
+      // Source is ParDo.. Check sink(s)
+      for (InputPortInfo sink : sinks) {
+        PTransform transform = sink.transform.getTransform();
+        if (transform instanceof ParDo.Bound) {
+          ParDo.Bound t = (ParDo.Bound) transform;
+          if (t.getSideInputs().size() > 0) {
+            loc = DAG.Locality.CONTAINER_LOCAL;
```

Comment:
What if side inputs are used by multiple transforms that are not part of a single fused chain? Then they should not be forced in the same container.